test/events_test.exs

```diff
@@ -1,3 +1,120 @@
+      raise "Got an event"
+    end
+  end
+
+  setup do
+    subscription_id = :crypto.strong_rand_bytes(16) |> Base.encode16()
+
+    on_exit fn ->
+      Events.unsubscribe(subscription_id)
+    end
+
+    {:ok, sub_id: subscription_id}
+  end
+
+  test "subscribing returns error if subscription with the same ID already exist", %{
+    sub_id: sub_id
+  } do
+    :ok = Events.subscribe(sub_id, [:some, :event], TestSubscriber, :echo_event)
+
+    assert :error = Events.subscribe(sub_id, [:some, :event], TestSubscriber, :echo_event)
+  end
+
+  test "subscribed mf is called when event exactly matches the subscription prefix", %{
+    sub_id: sub_id
+  } do
+    event = [:a, :test, :event]
+    config = %{send_to: self()}
+    value = 1
+    Events.subscribe(sub_id, event, TestSubscriber, :echo_event, config)
+
+    Events.emit(event, value)
+
+    assert_receive {:event, ^event, ^value, ^config}
+  end
+
+  test "subscribed mf is not called when event is a prefix of the subscription prefix", %{
+    sub_id: sub_id
+  } do
+    event = [:a, :test]
+    prefix = event ++ [:prefix]
+    config = %{send_to: self()}
+    value = 1
+    Events.subscribe(sub_id, prefix, TestSubscriber, :echo_event, config)
+
+    Events.emit(event, value)
+
+    assert {:messages, []} == Process.info(self(), :messages)
+  end
+
+  test "mf is unsubsribed when it fails", %{sub_id: sub_id} do
+    event = [:a, :test, :event]
+    Events.subscribe(sub_id, event, TestSubscriber, :raise_on_event)
+
+    Events.emit(event, 1)
+
+    assert [] == Events.list_subscriptions(event)
+  end
+
+  test "unsubcribed mf is not called when event is emitted", %{sub_id: sub_id} do
+    event = [:a, :test, :event]
+    Events.subscribe(sub_id, event, TestSubscriber, :echo_event, %{send_to: self()})
+
+    Events.unsubscribe(sub_id)
+    Events.emit(event, 1)
+
+    assert {:messages, []} == Process.info(self(), :messages)
+  end
+
+  test "unsubscribing returns error if subscription doesn't exist", %{
+    sub_id: sub_id
+  } do
+    assert :error = Events.unsubscribe(sub_id)
+  end
+
+  test "mf subscribed to event prefix is called when event is emitted", %{sub_id: sub_id} do
+    prefix = [:a, :test]
+    event = prefix ++ [:event]
+    config = %{send_to: self()}
+    value = 1
+    Events.subscribe(sub_id, prefix, TestSubscriber, :echo_event, config)
+
+    Events.emit(event, value)
+
+    assert_receive {:event, ^event, ^value, ^config}
+  end
+
+  test "subscriptions to event can be listed", %{sub_id: sub_id} do
+    event = [:a, :test, :event]
+    config = %{send_to: self()}
+    Events.subscribe(sub_id, event, TestSubscriber, :echo_event, config)
+
+    assert [{sub_id, event, TestSubscriber, :echo_event, config}] ==
+             Events.list_subscriptions(event)
+  end
+
+  test "subscriptions to event prefix can be listed", %{sub_id: sub_id} do
+    prefix1 = []
+    prefix2 = prefix1 ++ [:a]
+    prefix3 = prefix2 ++ [:test]
+    event = prefix3 ++ [:event]
```

Comment:
I personally believe that having prefixes hardcoded:
```
prefix1 = []
prefix2 = [:a]
prefix3 = [:a, :test]
event = [:a, :test, :event]
```
would make the test more readable.